src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -433,7 +433,13 @@ public void Abort(IOException abortReason)
                 return;
             }
 
-            AbortCore(abortReason);
+            AbortCore();
+
+            // Unblock the request body.
+            PoisonRequestBodyStream(new IOException(abortReason.Message, abortReason));
+            RequestBodyPipe.Writer.Complete(new IOException(abortReason.Message, abortReason));
```

Comment:
I don't think we really need to new up two exceptions here, since there should only be one reader accessing a single HTTP/2 request body at a time.

@Tratcher made a good point that it might be better to have the caller of Abort or ResetAndAbort clone the exceptions if needed. After all, it's the caller that knows if the exception is being used to complete multiple streams. As it is, this seems a little overly defensive, though certainly better than what's currently in master.

Reply:
Except its used by Request `PoisonRequestBodyStream` and Response `Writer.Complete` which can happen in parallel?

Reply:
Sure, but it's only a bug if it get's thrown/handled in parallel, right? There's only one reader for each stream who will either observe the exception via the poisoned request body stream OR the poisoned request body pipe.

The reader could only observe both exceptions it swallowed the exception from the pipe first and then tried reading again anyway thereby observing the the exception from the poisoned stream. This wouldn't cause parallel access to the exception though.

Reply:
The only obvious place I can see where it will keep adding to the exception stack trace is in `HttpRequestPipeReader` https://github.com/aspnet/AspNetCore/pull/11875 but I would have thought that would have shown up in the stack trace?

`Http2Connection` already creates a new exception for each `Http2Stream` https://github.com/aspnet/AspNetCore/blob/3f73a0fdcff217d061b994f3bd6528f1cd8caf2e/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L279-L282

However the stack trace would suggest either its happening in parallel or its chaining Write+Flush on Write+Flush on Write+Flush without ever waiting for any completion; so an earlier Flush would be active concurrently with a later Write?

Reply:
I really want to find the root of the problem. The exception set here should only be observed once or maybe repeatedly thrown to one caller if the caller happens swallow the exception. I really doubt that we're adding to the exception stack trace because we're throwing from both the poisoned Stream and PipeReader.

Reply:
Question is, do you think that could be an accurate the stacktrace?

Reply:
> Question is, do you think that could be an accurate the stacktrace?

I hope not, but I haven't ruled it out completely yet.

Reply:
Once #12081 is merged, TimingPipeFlusher will no longer be maintaining a flush Task queue. Instead ConcurrentQueue will be doing this job.

I think ConcurrentQueue stack traces will be easier to reason about because there is only ever one TCS at a time configured with TaskCreationOptions.RunContinuationsAsynchronously instead of an arbitrarily long linked list of Tasks being awaited.